### PR TITLE
Use typescript-eslint

### PR DESCRIPTION
**Pull request by davidpmccormick**
Ensuring TS is linted with 'recommended' settings.

I didn't want to go through the repo and lint all the components at the same time as adding this, but initial impressions are not too much will change. The main difference I can see is that `type` and `interface` declarations will have semicolons instead of commas.

If we want to go with this configuration, then it probably makes sense just to update the components as and when we touch those files, instead of doing a full-repo sweep?

### CODE REVIEW

**Review by jamesgorrie — COMMENTED** [review]
I think I would do a sweep and do it as part of this PR.

You could run the lint with autofix and you should get most of everything.

We should probably get all PRs in first - and any outstanding ones will have to deal with the conflicts.

**Comment by davidpmccormick** [comment]
> I think I would do a sweep and do it as part of this PR.

Right-oh.

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+const sharedRules = {
+  'no-multi-spaces': 'off',
+  'no-multi-str': 'off',
+  'no-mixed-operators': 'off',
+  'no-return-assign': 'off',
+  'react/react-in-jsx-scope': 'off',
+  'react-hooks/rules-of-hooks': 'error',
+  'prettier/prettier': 'error',
+};
+
```
Comment (jamesgorrie) [inline .eslintrc.js:28]: Is the `sharedX` stuff between flow and TS and we'll remove them once we're on one type system?

File: catalogue/webapp/components/Work/Work.tsx [hunk 2]
```diff
@@ -41,15 +41,15 @@ const ArchiveDetailsContainer = styled.div`
...
+    dataLayer: Record<string, unknown>[];
   }
 }
 
 type Props = {
   work: WorkType;
 };
 
-const Work = ({ work }: Props) => {
+const Work = ({ work }: Props): JSX.Element => {
```
Comment (jamesgorrie) [inline catalogue/webapp/components/Work/Work.tsx:52]: Should these be `FunctionComponent`s?

File: .eslintrc.js (hunk 1 above)
Comment (davidpmccormick) [reply .eslintrc.js:28]: yeah

File: catalogue/webapp/pages/images.tsx [hunk 3]
```diff
@@ -71,18 +79,22 @@ const ImagesPagination = ({
   </div>
 );
 
-const Images = ({ results, imagesRouteProps, apiProps }: Props) => {
+const Images = ({
+  results,
+  imagesRouteProps,
+  apiProps,
+}: Props): JSX.Element => {
```
Comment (jamesgorrie) [inline catalogue/webapp/pages/images.tsx:86]: `NextPage` type?

Comment (davidpmccormick) [reply catalogue/webapp/pages/images.tsx:86]: ah, possibly

File: common/views/components/BaseTabs/BaseTabs.tsx [hunk 4]
```diff
@@ -65,7 +65,7 @@ const Tabs = ({
   activeTabIndex,
   onTabClick,
   onTabChanged,
-}: Props) => {
+}: Props): JSX.Element => {
```
Comment (jamesgorrie) [inline common/views/components/BaseTabs/BaseTabs.tsx:68]: `FunctionComponent`?

File: common/views/components/BookImage/BookImage.tsx [hunk 5]
```diff
@@ -32,16 +31,18 @@ const Inner = styled('div').attrs({
 `;
 
 type Props = {
-  image: UiImageType,
-}
+  image: UiImageType;
+};
 
-const BookImage = ({ image }: Props) => {
+const BookImage = ({ image }: Props): JSX.Element => {
```
Comment (jamesgorrie) [inline common/views/components/BookImage/BookImage.tsx:37]: `FunctionComponent`?

File: common/views/components/ButtonInlineLink/ButtonInlineLink.tsx [hunk 6]
```diff
@@ -1,21 +1,24 @@
...
 type ButtonInlineLinkProps = ButtonInlineBaseProps & {
-  clickHandler?: (event: SyntheticEvent<HTMLAnchorElement>) => void,
-  link: {href: {pathname: string, query: string}, as: {pathname: string, query: string}} | string,
+  clickHandler?: (event: SyntheticEvent<HTMLAnchorElement>) => void;
+  link:
+    | {
+        href: { pathname: string; query: string };
+        as: { pathname: string; query: string };
+      }
+    | string;
```
Comment (jamesgorrie) [inline common/views/components/ButtonInlineLink/ButtonInlineLink.tsx:21]: These typed definitely exist somewhere else, but not for the linting PR

File: common/views/components/ButtonInlineLink/ButtonInlineLink.tsx [hunk 7]
```diff
@@ -26,7 +29,7 @@ const ButtonInlineLink = ({
   clickHandler,
   ariaControls,
   ariaExpanded,
-}: ButtonInlineLinkProps) => {
+}: ButtonInlineLinkProps): JSX.Element => {
```
Comment (jamesgorrie) [inline common/views/components/ButtonInlineLink/ButtonInlineLink.tsx:32]: `FunctionComponent`?

File: common/views/components/ButtonOutlinedLink/ButtonOutlinedLink.tsx [hunk 8]
```diff
@@ -26,8 +29,8 @@ const ButtonOutlinedLink = ({
   clickHandler,
   ariaControls,
   ariaExpanded,
-  isOnDark
-}: ButtonOutlinedLinkProps) => {
+  isOnDark,
+}: ButtonOutlinedLinkProps): JSX.Element => {
```
Comment (jamesgorrie) [inline common/views/components/ButtonOutlinedLink/ButtonOutlinedLink.tsx:33]: `FunctionComponent`?

File: common/views/components/ButtonSolidLink/ButtonSolidLink.tsx [hunk 9]
```diff
@@ -29,7 +28,7 @@ const ButtonSolidLink = ({
   ariaControls,
   ariaExpanded,
   isBig,
-}: ButtonSolidLinkProps) => {
+}: ButtonSolidLinkProps): JSX.Element => {
```
Comment (jamesgorrie) [inline common/views/components/ButtonSolidLink/ButtonSolidLink.tsx:31]: `FunctionComponent`?

File: common/views/components/CheckboxRadio/CheckboxRadio.tsx [hunk 10]
```diff
@@ -72,7 +72,12 @@ type CheckboxRadioProps = {
   value: string;
 };
 
-function CheckboxRadio({ id, text, type, ...inputProps }: CheckboxRadioProps) {
+function CheckboxRadio({
+  id,
+  text,
+  type,
+  ...inputProps
+}: CheckboxRadioProps): JSX.Element {
```
Comment (jamesgorrie) [inline common/views/components/CheckboxRadio/CheckboxRadio.tsx:80]: `FunctionComponent`?

File: common/views/components/CovidSafetyMeasure/CovidSafetyMeasure.tsx [hunk 11]
```diff
@@ -16,7 +16,12 @@ type Props = {
   isStacked?: boolean;
 };
 
-const CovidSafetyMeasure = ({ title, description, icon, isStacked }: Props) => {
+const CovidSafetyMeasure = ({
+  title,
+  description,
+  icon,
+  isStacked,
+}: Props): JSX.Element => {
```
Comment (jamesgorrie) [inline common/views/components/CovidSafetyMeasure/CovidSafetyMeasure.tsx:24]: `FunctionComponent`?

File: common/views/components/DropdownButton/DropdownButton.tsx [hunk 12]
```diff
@@ -72,7 +72,13 @@ type Props = {
...
 };
 
-const DropdownButton = ({ label, children, isInline, isOnDark, id }: Props) => {
+const DropdownButton = ({
+  label,
+  children,
+  isInline,
+  isOnDark,
+  id,
+}: Props): JSX.Element => {
```
Comment (jamesgorrie) [inline common/views/components/DropdownButton/DropdownButton.tsx:81]: `FunctionComponent`?

File: common/views/components/MediaObjectList/MediaObjectList.tsx [hunk 13]
```diff
@@ -1,14 +1,18 @@
 import { MediaObjectType } from '../../../model/media-object';
 import MediaObject from '../../components/MediaObject/MediaObject';
 
-const MediaObjectList = ({ items }: { items: Array<MediaObjectType> }) => {
+const MediaObjectList = ({
+  items,
+}: {
+  items: Array<MediaObjectType>;
```
Comment (jamesgorrie) [inline common/views/components/MediaObjectList/MediaObjectList.tsx:7]:
```suggestion
  items: MediaObjectType[];
```
👇 ?
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/array-type.md

**Review by jamesgorrie — COMMENTED** [review]
This is great! Just the one question around using `FunctionComponent` over `JSX.Element`.

File: catalogue/webapp/components/Work/Work.tsx (hunk 2 above)
Comment (davidpmccormick) [reply catalogue/webapp/components/Work/Work.tsx:52]: > A function component is of type `React.FunctionComponent` and it has an alias `React.FC` to keep things nice and short. It has one required property, a function, which will return a `ReactElement` or `null`. It has a few optional properties, such as `propTypes`, `contextTypes`, `defaultProps` and `displayName`.

From [SO](https://stackoverflow.com/a/57363438/590593).

So I think `const Work: FunctionComponent<Props> = ({ work }: Props): ReactElement<Props>` is probably the strictest correct typing.

Comment (jamesgorrie) [reply catalogue/webapp/components/Work/Work.tsx:52]: The return type is inferred so you don't have to put it there.

File: catalogue/webapp/pages/progress.tsx [hunk 14]
```diff
@@ -3,13 +3,14 @@ import { grid } from '@weco/common/utils/classnames';
...
 import Space from '@weco/common/views/components/styled/Space';
+import { ReactElement } from 'react';
 
 const title = "How we're improving search";
 const description =
   'We are working to make a more welcoming space where you' +
   'can discover more of what Wellcome Collection has to offer.';
 
-const ProgressPage = () => (
+const ProgressPage = (): ReactElement => (
```
Comment (jamesgorrie) [inline catalogue/webapp/pages/progress.tsx:13]:
```suggestion
const ProgressPage: NextPage = (): ReactElement => (
```

**Review by jamesgorrie — APPROVED** [review]

File: catalogue/webapp/components/Work/Work.tsx (hunk 2 above)
Comment (davidpmccormick) [reply catalogue/webapp/components/Work/Work.tsx]: Oh, only saw this after I'd added all the `ReactElements`. You're right, of course. Another pr to get rid of them, probably.

**Pull request merged**